### PR TITLE
[netcdf-c] Update to 4.8.1, revise features

### PR DESCRIPTION
**Pull request by dg0yt**
- #### What does your PR fix?  
  Updates netCDF-C.
  Decouples netcdf-4 feature from hdf5. (Hdf5 implies netcdf-4, but not vice versa.)
  Fixes hdf5 szip usage with static linkage (assuming it was broken already in previous versions).
  Exposes features for the new NCZarr support. (Needs for #21240 for building, but this should not hold back this PR.)
  Enables CI build with reduced feature set for uwp and arm64-windows.
  Enables using reduced set of dependencies (via explicit feature selection in manifest or with `vcpkg install`).

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  unchanged, no

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  yes

### CODE REVIEW

File: ports/netcdf-c/portfile.cmake [hunk 1]
```diff
@@ -19,22 +19,47 @@ file(REMOVE "${SOURCE_PATH}/cmake/modules/FindSZIP.cmake")
...
-        netcdf-4  CMAKE_DISABLE_FIND_PACKAGE_HDF5
-)
+        nczarr-zip    CMAKE_DISABLE_FIND_PACKAGE_Zip
+    )
+
+if(NOT ENABLE_DAP AND NOT ENABLE_NCZARR)
+    list(APPEND FEATURE_OPTIONS "-DCMAKE_DISABLE_FIND_PACKAGE_CURL=ON")
+endif()
+
+if(ENABLE_HDF5)
```
Comment (JackBoosY) [inline ports/netcdf-c/portfile.cmake:47]: Please note the macros `ENABLE_*` are not provided by `vcpkg_check_features`!
`vcpkg_check_features` can only set `ENABLE_*=ON` / `ENABLE_*=OFF` to `FEATURE_OPTIONS`.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/netcdf-c/portfile.cmake (hunk 1 above)
Comment (dg0yt) [reply ports/netcdf-c/portfile.cmake:47]: The macros *are provided*. YES `vcpkg_check_features` has side effects. I don't appreciate it, but many ports already rely on it, e.g. via `configure_file`.

Comment (dg0yt) [reply ports/netcdf-c/portfile.cmake:47]: Ping @JackBoosY. This has been discussed before in other PRs. Waiting for your response, not for "author reponse".

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
When building `pcl[*]:x86-windows`:
```
CMake Error at G:/21239/vcpkg/installed/x86-windows/share/AWSSDK/AWSSDKConfig.cmake:99 (message):
  AWS SDK for C++ is missing, please install it first
```

**Comment by dg0yt** [comment]
> CMake Error at G:/21239/vcpkg/installed/x86-windows/share/AWSSDK/AWSSDKConfig.cmake:99 (message):
>   AWS SDK for C++ is missing, please install it first

@JackBoosY See the initial post: I provided another PR, and it is already merged.

**Comment by JackBoosY** [comment]
Another issue:
```
[226/233] cmd.exe /C "cd . && G:\21239\vcpkg\downloads\tools\cmake-3.21.1-windows\cmake-3.21.1-windows-i386\bin\cmake.exe -E vs_link_exe --intdir=ncdump\CMakeFiles\nc4print.dir --rc="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\rc.exe" --mt="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\mt.exe" --manifests  -- "C:\PROGRA~2\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\Hostx64\x64\link.exe"  ncdump\CMakeFiles\nc4print.dir\nc4print.c.obj ncdump\CMakeFiles\nc4print.dir\nc4printer.c.obj  /out:ncdump\nc4print.exe /implib:ncdump\nc4print.lib /pdb:ncdump\nc4print.pdb /version:0.0 /machine:x64 /LARGEADDRESSAWARE /STACK:40000000 /nologo    /debug /INCREMENTAL /subsystem:console -LIBPATH:G:\21239\vcpkg\installed\x64-windows-static\lib liblib\netcdf.lib  G:\21239\vcpkg\installed\x64-windows-static\debug\lib\libhdf5_D.lib  G:\21239\vcpkg\installed\x64-windows-static\debug\lib\libhdf5_hl_D.lib  G:\21239\vcpkg\installed\x64-windows-static\debug\lib\libcurl-d.lib  G:\21239\vcpkg\installed\x64-windows-static\lib\zip.lib  G:\21239\vcpkg\installed\x64-windows-static\debug\lib\libhdf5_D.lib  G:\21239\vcpkg\installed\x64-windows-static\debug\lib\zlibd.lib  G:\21239\vcpkg\installed\x64-windows-static\lib\libszip.lib  winmm.lib  G:\21239\vcpkg\installed\x64-windows-static\debug\lib\zlibd.lib  advapi32.lib  crypt32.lib  aws-cpp-sdk-s3.lib  aws-cpp-sdk-core.lib  aws-crt-cpp.lib  aws-c-auth.lib  aws-c-cal.lib  aws-c-common.lib  aws-c-compression.lib  aws-c-event-stream.lib  aws-c-http.lib  aws-c-io.lib  aws-c-mqtt.lib  aws-c-s3.lib  aws-checksums.lib  Userenv.lib  version.lib  ws2_32.lib  Bcrypt.lib  Wininet.lib  winhttp.lib  G:\21239\vcpkg\installed\x64-windows-static\lib\zip.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
FAILED: ncdump/nc4print.exe 
cmd.exe /C "cd . && G:\21239\vcpkg\downloads\tools\cmake-3.21.1-windows\cmake-3.21.1-windows-i386\bin\cmake.exe -E vs_link_exe --intdir=ncdump\CMakeFiles\nc4print.dir --rc="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\rc.exe" --mt="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\mt.exe" --manifests  -- "C:\PROGRA~2\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\Hostx64\x64\link.exe"  ncdump\CMakeFiles\nc4print.dir\nc4print.c.obj ncdump\CMakeFiles\nc4print.dir\nc4printer.c.obj  /out:ncdump\nc4print.exe /implib:ncdump\nc4print.lib /pdb:ncdump\nc4print.pdb /version:0.0 /machine:x64 /LARGEADDRESSAWARE /STACK:40000000 /nologo    /debug /INCREMENTAL /subsystem:console -LIBPATH:G:\21239\vcpkg\installed\x64-windows-static\lib liblib\netcdf.lib  G:\21239\vcpkg\installed\x64-windows-static\debug\lib\libhdf5_D.lib  G:\21239\vcpkg\installed\x64-windows-static\debug\lib\libhdf5_hl_D.lib  G:\21239\vcpkg\installed\x64-windows-static\debug\lib\libcurl-d.lib  G:\21239\vcpkg\installed\x64-windows-static\lib\zip.lib  G:\21239\vcpkg\installed\x64-windows-static\debug\lib\libhdf5_D.lib  G:\21239\vcpkg\installed\x64-windows-static\debug\lib\zlibd.lib  G:\21239\vcpkg\installed\x64-windows-static\lib\libszip.lib  winmm.lib  G:\21239\vcpkg\installed\x64-windows-static\debug\lib\zlibd.lib  advapi32.lib  crypt32.lib  aws-cpp-sdk-s3.lib  aws-cpp-sdk-core.lib  aws-crt-cpp.lib  aws-c-auth.lib  aws-c-cal.lib  aws-c-common.lib  aws-c-compression.lib  aws-c-event-stream.lib  aws-c-http.lib  aws-c-io.lib  aws-c-mqtt.lib  aws-c-s3.lib  aws-checksums.lib  Userenv.lib  version.lib  ws2_32.lib  Bcrypt.lib  Wininet.lib  winhttp.lib  G:\21239\vcpkg\installed\x64-windows-static\lib\zip.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
LINK Pass 1: command "C:\PROGRA~2\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\Hostx64\x64\link.exe ncdump\CMakeFiles\nc4print.dir\nc4print.c.obj ncdump\CMakeFiles\nc4print.dir\nc4printer.c.obj /out:ncdump\nc4print.exe /implib:ncdump\nc4print.lib /pdb:ncdump\nc4print.pdb /version:0.0 /machine:x64 /LARGEADDRESSAWARE /STACK:40000000 /nologo /debug /INCREMENTAL /subsystem:console -LIBPATH:G:\21239\vcpkg\installed\x64-windows-static\lib liblib\netcdf.lib G:\21239\vcpkg\installed\x64-windows-static\debug\lib\libhdf5_D.lib G:\21239\vcpkg\installed\x64-windows-static\debug\lib\libhdf5_hl_D.lib G:\21239\vcpkg\installed\x64-windows-static\debug\lib\libcurl-d.lib G:\21239\vcpkg\installed\x64-windows-static\lib\zip.lib G:\21239\vcpkg\installed\x64-windows-static\debug\lib\libhdf5_D.lib G:\21239\vcpkg\installed\x64-windows-static\debug\lib\zlibd.lib G:\21239\vcpkg\installed\x64-windows-static\lib\libszip.lib winmm.lib G:\21239\vcpkg\installed\x64-windows-static\debug\lib\zlibd.lib advapi32.lib crypt32.lib aws-cpp-sdk-s3.lib aws-cpp-sdk-core.lib aws-crt-cpp.lib aws-c-auth.lib aws-c-cal.lib aws-c-common.lib aws-c-compression.lib aws-c-event-stream.lib aws-c-http.lib aws-c-io.lib aws-c-mqtt.lib aws-c-s3.lib aws-checksums.lib Userenv.lib version.lib ws2_32.lib Bcrypt.lib Wininet.lib winhttp.lib G:\21239\vcpkg\installed\x64-windows-static\lib\zip.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:ncdump\CMakeFiles\nc4print.dir/intermediate.manifest ncdump\CMakeFiles\nc4print.dir/manifest.res" failed (exit code 1120) with the following output:
aws-cpp-sdk-s3.lib(ub_S3.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in netcdf.lib(zs3sdk.cpp.obj)
aws-cpp-sdk-s3.lib(ub_S3.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MTd_StaticDebug' in netcdf.lib(zs3sdk.cpp.obj)
aws-cpp-sdk-core.lib(ub_core.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in netcdf.lib(zs3sdk.cpp.obj)
aws-cpp-sdk-core.lib(ub_core.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MTd_StaticDebug' in netcdf.lib(zs3sdk.cpp.obj)
aws-cpp-sdk-core.lib(Aws.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in netcdf.lib(zs3sdk.cpp.obj)
```

**Comment by dg0yt** [comment]
> `mismatch detected for '_ITERATOR_DEBUG_LEVEL'`

ABI mismatch, good catch. But what is the desired iterator debug level in vcpkg? Only one port can be right.

**Comment by dg0yt** [comment]
> `[226/233] .. /debug ... -LIBPATH:G:\21239\vcpkg\installed\x64-windows-static\lib liblib\netcdf.lib  ... G:\21239\vcpkg\installed\x64-windows-static\debug\lib\libcurl-d.lib  G:\21239\vcpkg\installed\x64-windows-static\lib\zip.lib  G:\21239\vcpkg\installed\x64-windows-static\debug\lib\libhdf5_D.lib ... G:\21239\vcpkg\installed\x64-windows-static\lib\libszip.lib  ...`

... and here is a strange mix of debug and release 💥  We need to check the imported configurations.

**Comment by dg0yt** [comment]
>  here is a strange mix of debug and release

This configuration comes from `installed/x64-linux/share/aws-cpp-sdk-core/aws-cpp-sdk-core-targets.cmake`:  

`INTERFACE_LINK_LIBRARIES "\$<LINK_ONLY:pthread>;\$<\$<NOT:\$<CONFIG:DEBUG>>:${_IMPORT_PREFIX}/lib/libcurl.a>;\$<\$<CONFIG:DEBUG>:${_IMPORT_PREFIX}/debug/lib/libcurl-d.a>;\$<LINK_ONLY:dl>;\$<LINK_ONLY:-lpthread>;\$<LINK_ONLY:OpenSSL::SSL>;\$<LINK_ONLY:OpenSSL::Crypto>;\$<LINK_ONLY:ZLIB::ZLIB>;${_IMPORT_PREFIX}/lib/libssl.a;${_IMPORT_PREFIX}/lib/libcrypto.a;\$<LINK_ONLY:-lpthread>;\$<LINK_ONLY:dl>;\$<LINK_ONLY:dl>;\$<LINK_ONLY:-lpthread>;${_IMPORT_PREFIX}/lib/libz.a;AWS::aws-crt-cpp"`

Related: https://github.com/microsoft/vcpkg/pull/20289#issuecomment-962888771

**Comment by dg0yt** [comment]
aws-cpp-sdk is a difficult citizen in vcpkg. It is not ready for vcpkg-style debug+release. It exports dependencies incompletely (missing: s2n, stdc++) and not in the right order for static linkage.

**Comment by JackBoosY** [comment]
@dg0yt Can you please report that to https://github.com/aws/aws-sdk-cpp?

**Comment by dg0yt** [comment]
> @dg0yt Can you please report that to https://github.com/aws/aws-sdk-cpp?

I think we need to  make port aws-sdk-cpp etc. up-to-date first.

I would suggest to remove feature nczarr-s3 from this port-version of netcdf-c. It needs more work in vcpkg and in upstreams.
I can keep the current state in a separate branch of my vcpkg fork.

**Comment by JackBoosY** [comment]
I will restart x64-osx test after the other tests finished.

**Comment by dg0yt** [comment]
You may re-run CI now. x64-linux is blocked for no obvious reason.

**Comment by JackBoosY** [comment]
It seems that the feature changes fails to run.

**Comment by dg0yt** [comment]
Strange. The last commit didn't change a bit of port netcdf-c, and the debug run of `vcpkg ci` wasn`t used, even on osx. Reverting the last commit (as it is unrelated).

**Comment by JackBoosY** [comment]
All features are tested successfully on x86-windows and x64-windows-static.

**Pull request merged**